python/cuml/feature_extraction/text.py

```diff
@@ -0,0 +1,603 @@
+    non_alpha = characters.str.extract(r'([^\w])', expand=False).dropna()
+    return non_alpha.tolist() + ['\n']
+
+
+def _preprocess(doc, lower=False, remove_non_words=False, delimiter=' '):
+    """Chain together an optional series of text preprocessing steps to
+    apply to a document.
+    Parameters
+    ----------
+    doc: nvstrings
+        The string to preprocess
+    lower: bool
+        Whether to use str.lower to lowercase all of the text
+    remove_non_words: bool
+        Whether or not to remove non-alphanumeric characters.
+    Returns
+    -------
+    doc: nvstrings
+        preprocessed string
+    """
+    if lower:
+        doc = doc.str.lower()
+    if remove_non_words:
+        non_alpha = _get_non_alphanumeric_characters(doc)
+        delimiter_code = ord(delimiter)
+        translation_table = {ord(char): delimiter_code for char in non_alpha}
+        doc = doc.str.translate(translation_table)
+    return doc
+
+
+class _VectorizerMixin:
+    """Provides common code for text vectorizers (tokenization logic)."""
+    def _remove_stop_words(self, doc):
+        """Remove stop words only if needed."""
+        if self.analyzer == 'word' and self.stop_words is not None:
+            stop_words = Series(self._get_stop_words())
+            doc = doc.str.replace_tokens(stop_words,
+                                         replacements=self.delimiter,
+                                         delimiter=self.delimiter)
+        return doc
+
+    def build_preprocessor(self):
+        """Return a function to preprocess the text before tokenization.
+
+        If analyzer == 'word' and stop_words is not None, stop words are
+        removed from the input documents after preprocessing.
+
+        Returns
+        -------
+        preprocessor: callable
+              A function to preprocess the text before tokenization.
+        """
+        if self.preprocessor is not None:
+            preprocess = self.preprocessor
+        else:
+            preprocess = partial(_preprocess, lower=self.lowercase,
+                                 remove_non_words=self.analyzer == 'word',
+                                 delimiter=self.delimiter)
+        return lambda doc: self._remove_stop_words(preprocess(doc))
+
+    def _get_stop_words(self):
+        """Build or fetch the effective stop words list.
+        Returns
+        -------
+        stop_words: list or None
+                A list of stop words.
+        """
+        if self.stop_words == "english":
+            return list(ENGLISH_STOP_WORDS)
+        elif isinstance(self.stop_words, str):
+            raise ValueError("not a built-in stop list: %s" % self.stop_words)
+        elif self.stop_words is None:
+            return None
+        else:  # assume it's a collection
+            return list(self.stop_words)
+
+    def get_char_ngrams(self, n, str_series, doc_id_sr):
+        """
+        Handles ngram generation for characters analyzers.
```

Comment:
with a simple name like 'n' it would be helpful to have the docstring explain this parameter. I think it's max ngram length?

Reply:
Good point, it's not the max but the current ngram size. I will clarify this in the code.